### PR TITLE
Dev branch  user space

### PR DESCRIPTION
**Pull request by EricGebhart**
rewrite/refactor of user space, some additional keymaps.

## Description

A couple of years worth of user space changes.  a map for crkbd and kinesis. Updates to 
the keymaps as reflected by the user space.  A mod layer, home row mods, better keymap organisation,
oled support.   Much better organization of user.

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [x] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Review by drashna — CHANGES_REQUESTED** [review]

File: keyboards/crkbd/keymaps/ericgebhart/config.h [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+#define SPLIT_USB_DETECT
+
+//#define TAPPING_FORCE_HOLD
+
+// #define TAPPING_TERM 100
+
+// #define RGBLIGHT_ENABLE
+#ifdef RGBLIGHT_ENABLE
+#undef RGBLED_NUM
+#define RGBLIGHT_ANIMATIONS
```
Comment (drashna) [inline keyboards/crkbd/keymaps/ericgebhart/config.h:40]: Could you explicitly define which animations are enabled? 
```suggestion
#define RGBLIGHT_EFFECT_BREATHING
#define RGBLIGHT_EFFECT_RAINBOW_MOOD
#define RGBLIGHT_EFFECT_RAINBOW_SWIRL
#define RGBLIGHT_EFFECT_SNAKE
#define RGBLIGHT_EFFECT_KNIGHT
#define RGBLIGHT_EFFECT_CHRISTMAS
#define RGBLIGHT_EFFECT_STATIC_GRADIENT
#define RGBLIGHT_EFFECT_RGB_TEST
#define RGBLIGHT_EFFECT_ALTERNATING
#define RGBLIGHT_EFFECT_TWINKLE
```

File: keyboards/crkbd/keymaps/ericgebhart/keymap.c [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+#include "ericgebhart.h"
+
+extern uint8_t is_master;
+
+#ifdef RGBLIGHT_ENABLE
+// Following line allows macro to read current RGB settings
+extern rgblight_config_t rgblight_config;
+#endif
```
Comment (drashna) [inline keyboards/crkbd/keymaps/ericgebhart/keymap.c:24]:
```suggestion
```

File: keyboards/crkbd/keymaps/ericgebhart/rules.mk [hunk 3]
```diff
@@ -0,0 +1,2 @@
+OLED_ENABLE = yes
+OLED_DRIVER = SSD1306
```
Comment (drashna) [inline keyboards/crkbd/keymaps/ericgebhart/rules.mk:2]: This is the default, actually.

**Review by drashna — CHANGES_REQUESTED** [review]

File: keyboards/crkbd/keymaps/ericgebhart/config.h (hunk 1 above)
Comment (EricGebhart) [reply keyboards/crkbd/keymaps/ericgebhart/config.h:40]: I should remove them, I don't use them really, I'm out of memory on 
my AVRs. I trimmed it down to the only thing that actually matters I think.

File: keyboards/crkbd/keymaps/ericgebhart/keymap.c (hunk 2 above)
Comment (EricGebhart) [reply keyboards/crkbd/keymaps/ericgebhart/keymap.c]: removed.

File: keyboards/crkbd/keymaps/ericgebhart/rules.mk (hunk 3 above)
Comment (EricGebhart) [reply keyboards/crkbd/keymaps/ericgebhart/rules.mk:2]: Sometimes I turn it off. I guess it doesn't really matter.

Comment (EricGebhart) [reply keyboards/crkbd/keymaps/ericgebhart/rules.mk]: removed it.

**Review by drashna — APPROVED** [review]

File: users/ericgebhart/process_records.c [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+    }
+  }
+}
+
+// Defines actions for my global custom keycodes. Defined in ericgebhart.h file
+// Then runs the _keymap's record handier if not processed here
+bool process_record_user(uint16_t keycode, keyrecord_t *record) {
+  // If console is enabled, it will print the matrix position and status of each key pressed
+
+#ifdef OLED_DRIVER_ENABLE
```
Comment (daskygit) [inline users/ericgebhart/process_records.c:61]: Not sure if you want this as `CUSTOM_OLED_DRIVER_CODE` as per the define in your rules or the default `OLED_ENABLE`

File: keyboards/montsinger/rebound/rev4/keymaps/ericgebhart/rules.mk [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+COMMAND_ENABLE = yes        # Commands for debug and configuration
+NKRO_ENABLE = yes           # Nkey Rollover - if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
+BACKLIGHT_ENABLE = no       # Enable keyboard backlight functionality
+MIDI_ENABLE = no            # MIDI controls
+AUDIO_ENABLE = no           # Audio output on port C6
+UNICODE_ENABLE = no         # Unicode
+BLUETOOTH_ENABLE = no       # Enable Bluetooth with the Adafruit EZ-Key HID
+RGBLIGHT_ENABLE = no        # Enable WS2812 RGB underlight.
+SLEEP_LED_ENABLE = no       # Breathing sleep LED during USB suspend
+TAP_DANCE_ENABLE = yes      # Enable the tap dance feature.
```
Comment (daskygit) [inline keyboards/montsinger/rebound/rev4/keymaps/ericgebhart/rules.mk:18]:
```suggestion
# Build Options
#   change yes to no to disable
#
BOOTMAGIC_ENABLE = no       # Enable Bootmagic Lite
MOUSEKEY_ENABLE = yes       # Mouse keys
EXTRAKEY_ENABLE = yes       # Audio control and System control
CONSOLE_ENABLE = no         # Console for debug
COMMAND_ENABLE = yes        # Commands for debug and configuration
NKRO_ENABLE = yes           # Enable N-Key Rollover
BACKLIGHT_ENABLE = no       # Enable keyboard backlight functionality
MIDI_ENABLE = no            # MIDI controls
AUDIO_ENABLE = no           # Audio output
UNICODE_ENABLE = no         # Unicode
RGBLIGHT_ENABLE = no        # Enable WS2812 RGB underlight.
SLEEP_LED_ENABLE = no       # Breathing sleep LED during USB suspend
TAP_DANCE_ENABLE = yes      # Enable the tap dance feature.
```

A lot of these options are default for the keyboard you should consider removing them.

File: keyboards/kinesis/keymaps/ericgebhart/rules.mk [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+CONSOLE_ENABLE = no         # Console for debug
+COMMAND_ENABLE = yes        # Commands for debug and configuration
+NKRO_ENABLE = yes           # Nkey Rollover - if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
+BACKLIGHT_ENABLE = no       # Enable keyboard backlight functionality
+MIDI_ENABLE = no            # MIDI controls
+AUDIO_ENABLE = no           # Audio output on port C6
+UNICODE_ENABLE = no         # Unicode
+BLUETOOTH_ENABLE = no       # Enable Bluetooth with the Adafruit EZ-Key HID
+RGBLIGHT_ENABLE = no        # Enable WS2812 RGB underlight. 
+SLEEP_LED_ENABLE = no       # Breathing sleep LED during USB suspend
```
Comment (daskygit) [inline keyboards/kinesis/keymaps/ericgebhart/rules.mk:17]:
```suggestion
# Build Options
#   change yes to no to disable
#
BOOTMAGIC_ENABLE = no       # Enable Bootmagic Lite
MOUSEKEY_ENABLE = yes       # Mouse keys
EXTRAKEY_ENABLE = yes       # Audio control and System control
CONSOLE_ENABLE = no         # Console for debug
COMMAND_ENABLE = yes        # Commands for debug and configuration
NKRO_ENABLE = yes           # Enable N-Key Rollover
BACKLIGHT_ENABLE = no       # Enable keyboard backlight functionality
MIDI_ENABLE = no            # MIDI controls
AUDIO_ENABLE = no           # Audio output
UNICODE_ENABLE = no         # Unicode
RGBLIGHT_ENABLE = no        # Enable WS2812 RGB underlight. 
SLEEP_LED_ENABLE = no       # Breathing sleep LED during USB suspend
```
A lot of these options are default for the keyboard you should consider removing them.

**Review by daskygit — CHANGES_REQUESTED** [review]

File: users/ericgebhart/rules.mk [hunk 7]
```diff
@@ -1,14 +1,38 @@
...
+
+CUSTOM_OLED_DRIVER ?= yes
+ifeq ($(strip $(OLED_ENABLE)), yes)
+    ifeq ($(strip $(CUSTOM_OLED_DRIVER)), yes)
+        SRC += $(USER_PATH)/oled_stuff.c
+        OPT_DEFS += -DCUSTOM_OLED_DRIVER_CODE
+    endif
+endif
+
+EXTRAFLAGS += -flto
```
Comment (daskygit) [inline users/ericgebhart/rules.mk:17]:
```suggestion
```
This is already done by `LTO_ENABLE = yes`

**Review by daskygit — APPROVED** [review]

**Comment by drashna** [comment]
Thanks!

**Pull request merged**